### PR TITLE
introduce caching

### PR DESCRIPTION
**Pull request by garak**
See #61 

Note that I had to raise requirements to Symfony 2.4 to be able to use expression language. This is not a problem, since Symfony 2.3 is out of support and a minor release of this bundle (including this PR) will respect semver.

### CODE REVIEW

File: README.markdown [hunk 1]
```diff
@@ -94,3 +94,24 @@ Or with parameters:
 For more information [look at the gravatar implementation pages][gravatar].
 
 [gravatar]: http://en.gravatar.com/site/implement/
+
+Caching
+-------
+
+You can cache calls to Gravatar API, using [DoctrineCahce](https://symfony.com/doc/current/bundles/DoctrineCacheBundle/index.html).
```
Comment (gharlan) [inline README.markdown]: typo (DoctrineCahce)

**Comment by gharlan** [comment]
> Wouldn't it be better to use symfony/cache component instead of doctrine/cache?

see also last comment here: https://github.com/henrikbjorn/GravatarBundle/pull/49#issuecomment-362212717

**Comment by garak** [comment]
I agree that symfony/cache would be better, but it would require un upgrade of requirements, since it's available only since symfony 3

**Comment by alister** [comment]
I'm not sure why caching would be useful - the only external call to an external service is in `GravatarApi::exists()`. All the rest are effectively string manipulation on the result of running `md5($email)`.

**Comment by garak** [comment]
@alister that call is repeated endless times, even if it's needed just once.
Typical example: you display a gravatar for your logged user in the navigation bar of your site. Without cache, every single page does an external call to gravatar service. With cache, only one call is done (inside timeframe of your cache)

**Comment by alister** [comment]
Yes, but you could cache that fact within your own code, or 1) have Gravatar return the auto-generated image (identicon, monster, retro, mystery-man, etc), 2) provide your own local default image URL (the third parameter to the `{{gravatar(...)}}` / `getUrl()`), or 3) have it return a 404 to be caught by Javascript and swap in a default.

All the other calls to the cache may be taking more time and effort than fetching the underlying calculations does, plus the added complexity/overhead of the cache.

**Comment by alister** [comment]
It's a library for getting URLs for Gravatar images and profiles. Depending on what you need to do, as per my last message you also have four other ways to act if the email address supplied does not have a customised gravatar image.

What is in your `{% else %}` block after the `{% if gravatar_exists(email) %}...`, or `if (GravatarApi::exists(...)) {`?

**Comment by alister** [comment]
What you want to do is already solved with the functionality currently in the bundle.

```
# config.yml
ornicar_gravatar:
    default: mm
```
or `{{ gravatar(profile.user.emailCanonical,200,'g','mm') }}`

If there is a specific image matching the email hash, it will show that, if not, it shows the 'Mystery Man' image.

If you specifically want the Font-awesome icon (or something similar), you can [make a .png of it](http://fa2png.io/r/font-awesome/user-circle/), and then supply the URL as the default instead of the 'mm' [gravatar.com/avatar/hash&d=*URL*](https://secure.gravatar.com/avatar/ac62d0bap5u340543058039458cd2?s=100&r=g&d=http%3A%2F%2Ffa2png.io%2Fmedia%2Ficons%2Ffont-awesome%2F4-7-0%2Fuser-circle%2F256%2F0%2F007dff_none.png)

**Comment by garak** [comment]
@alister unfortunately that's not enough. Still, each time `gravatar` function is called, a call to gravatar server, that needs to pass the whole Internet, is made.
Did you notice that I had to completely exclude such call in environment test? That's because my test suite gone from 20 seconds to 60

**Comment by alister** [comment]
I think you have some other issues - the `{{gravatar(...)}}` call in twig, or the underlying `GravatarApi::getUrl()` doesn't make any external calls - the only place in the entire bundle that talks to the outside world is the `GravatarApi::exists()` - the rest of the twig functions just return a string, based on the given email, and some other optional/default values.

The only way it would fetch from the URLs it returns (images, or profiles), is if you set it to - possibly via a full-browser test with Selenium (but not easily with the built-in Symfony client/crawler test).  Adding the `{% if app.environment != 'test' }}` would even do away with having the html img-tag from being output, if you wanted to stop that from showing up in the output html.

**Comment by henrikbjorn** [comment]
Hey @garak In the example of your code posted above you use `gravatar_exists` to check for the existance of a gravatar email. it is correct this makes a request to the gravatar servers.

What @alister is suggesting is to remove the use of `gravatar_exists` and instead rely on feature of gravatar where you can use a image as a fallback. He demonstrated this here https://github.com/henrikbjorn/GravatarBundle/pull/65#issuecomment-375815521

If your test is the issue then i will suggest you mock the GravatarApi service for yours tests which would remove the issue completely.

Caching will not help your case.

**Pull request closed without merging**